### PR TITLE
add Federated SVD++ model

### PR DESCRIPTION
**Pull request by liucongwust**
Signed-off-by: tonyliu <liucongwust@gmail.com>

Fixes  ISSUE#xxx

Changes:

1. add federated svd++ model (FedSVD++)

2. add examples of FedSVD++

### CODE REVIEW

File: federatedrec/param/svdpp_param.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        Required when using encrypted version HomoLR. Since multiple batch updating coefficient may cause
+        overflow error. The model need to be re-encrypt for every several batches. Please be careful when setting
+        this parameter. Too large batches may cause training failure.
+
+    aggregate_iters : int, default: 1
+        Indicate how many iterations are aggregated once.
+
+    """
+
+    def __init__(self, optimizer='sgd',
```
Comment (mgqa34) [inline federatedrec/param/svdpp_param.py]: When using tf's optimizer, it will be case sensitive. The default value optimizer='sgd' may raise some error?

File: federatedrec/svdpp/README.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+<div style="text-align:center", align=center>
+<img src="./images/fig6.png" alt="samples" width="443" height="60" /><br/>
+</div>
+
+## Features:
+
+1. L1 & L2 regularization
+2. Mini-batch mechanism
+3. Five optimization methods:
```
Comment (mgqa34) [inline federatedrec/svdpp/README.md]: Same here, the method's , "sgd", "nesterov_momentum_sgd" should be ajust.

File: federatedml/conf/default_runtime_conf/svdpp_param.json [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  "batch_size": 320,
+  "init_param": {
+    "init_method": "random_normal",
+    "embed_size": 20
+  },
+  "predict_param": {
+    "with_proba": true,
+    "threshold": 0.5
+  },
+  "encrypt_param": {
```
Comment (mgqa34) [inline federatedml/conf/default_runtime_conf/svdpp_param.json]: Encrypt_param does not use ?

File: federatedml/conf/default_runtime_conf/svdpp_param.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    "method": "Paillier",
+    "key_length": 1024
+  },
+  "encrypted_mode_calculator_param": {
```
Comment (mgqa34) [inline federatedml/conf/default_runtime_conf/svdpp_param.json]: Same with above, unused parameter

File: federatedrec/param/svdpp_param.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        self.init_method = init_method
+
+
+class SVDppParam(BaseParam):
+    """
+    Parameters used for Logistic Regression both for Homo mode or Hetero mode.
+
+    Parameters
+    ----------
+    optimizer : str, 'SGD', 'RMSprop', 'Adam' or 'Adagrad', default: 'sgd'
```
Comment (mgqa34) [inline federatedrec/param/svdpp_param.py]: default: 'SGD'

File: federatedrec/param/svdpp_param.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+            return param
+        else:
+            raise ValueError(f"invalid metrics type: {type(param)}")
+
+
+class HeteroSVDppParam(SVDppParam):
+    """
+    Parameters
+    ----------
+    re_encrypt_batches : int, default: 2
```
Comment (zazd) [inline federatedrec/param/svdpp_param.py]: The parameter annotation may be updated?

File: federatedrec/svdpp/hetero_svdpp/hetero_svdpp_client.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+        self._model = None
+
+    def _init_model(self, param):
+        super()._init_model(param)
+
+        self.batch_size = param.batch_size
+        self.aggregate_every_n_epoch = 1
+        self.optimizer = param.optimizer
+        self.loss = param.loss
```
Comment (zazd) [inline federatedrec/svdpp/hetero_svdpp/hetero_svdpp_client.py]: can parameter "loss" be set before training or be saved after training?

File: federatedrec/svdpp/README.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
+# Federated Singular Value Decomposition++
+
+SVD++ is a  supervised learning approach that decompose a matrix into a product of matrices. It does matrix factorization via following formula:
+
+<div style="text-align:center", align=center>
+<img src="./images/fig1.png" alt="samples" width="341" height="55" /><br/>
+</div>
+
+Different to SVD, SVD++ incoporate imlicit feedback of users.
```
Comment (zazd) [inline federatedrec/svdpp/README.md]: some word misspelling like "incoporate" ""imlicit" "intutively" ?

File: federatedrec/README.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+- Corresponding module name: HeteroSVD
+- Data Input: Input DTable of user-item rating matrix data.
+- Model Output: Hetero SVD model.
+
+
+##### 4. [Hetero SVD++](./svdpp/README.md)
+Build hetero SVD module through multiple parties.
+
+- Corresponding module name: HeteroSVD
```
Comment (zazd) [inline federatedrec/README.md]: This should be HeteroSVD++ or HeteroSVDPP?

File: federatedrec/svdpp/hetero_svdpp/hetero_svdpp_client.py (hunk 7 above)
Comment (liucongwust) [reply federatedrec/svdpp/hetero_svdpp/hetero_svdpp_client.py]: loss if type of String, used to build model. Keras model use this parameter to choose the corresponding loss function, which can be saved after training.
